compiler/rustc_const_eval/src/interpret/place.rs

```diff
@@ -869,8 +870,17 @@ where
             Ok(src_val) => {
                 assert!(!src.layout.is_unsized(), "cannot have unsized immediates");
                 // Yay, we got a value that we can write directly.
-                // FIXME: Add a check to make sure that if `src` is indirect,
-                // it does not overlap with `dest`.
+                // Then make sure `src` does not overlap with `dest`.
+                if let Operand::Indirect(src_mplace) = src.deref()
+                    && let Place::Ptr(dest_mplace) = dest.deref()
+                    && let Ok((src_id, src_offset, _)) = self.ptr_try_get_alloc_id(src_mplace.ptr)
+                    && let Ok((dest_id, dest_offset, _)) = self.ptr_try_get_alloc_id(dest_mplace.ptr)
+                {
+                    let size = src_val.layout.size;
+                    if Self::check_ptr_overlap(src_id, src_offset, dest_id, dest_offset, size) {
+                        throw_ub_format!("copy_nonoverlapping called on overlapping ranges")
```

Comment:
This is not the copy_nonoverlapping intrinsic, this is the assignment operator.

Also, that condition up there is enormously complicated and looks like it just unwraps all abstractions to get some deep implementation detail out. What is actually conceptually happening here, at a high level?

Reply:
I think the right way to fix this FIXME is what we are discussing in https://github.com/rust-lang/rust/issues/68364. I don't think digging through all abstractions here is a good way to guarantee anything, since it is basically impossible to figure out what is being guaranteed. It is way too easy for any of these conditions to not be met, and then we gain nothing.

Reply:
So, I think we shouldn't do this part of the PR. Sorry for the misleading FIXME. Generally many of the FIXME in the code are old and our thinking might have changed since the comment was written -- so it is a good idea to ask people what their current thoughts on this are before working on a FIXME.

Here, I think the comment should just be changed to explicitly reference https://github.com/rust-lang/rust/issues/68364.

Reply:
That's all right. Thanks!